SourceCodeGenerator: Add support for enum source code generation

xC0dex opened the pull request:
## Description
Currently, the extension method `ToDescriptionString()` for enums uses reflection to get the value of the `[Description]` attribute. 
This PR adds an implementation of `IIncrementalGenerator` to generate this extension method during the build. This improves general performance, especially in rendering. The following benchmark shows the performance boost:
|                 Method |       Mean |      Error |     StdDev | Ratio | Allocated | Alloc Ratio |
|----------------------- |-----------:|-----------:|-----------:|------:|----------:|------------:|
| ToDescriptionStringOld | 851.880 ns |  8.3776 ns |  7.8364 ns | 1.000 |     264 B |        1.00 |
| ToDescriptionStringNew |   1.686 ns |  0.0253 ns |  0.0237 ns | 0.002 |         - |        0.00 |

The reflection-based implementation is marked as obsolete, but still can be used. Now there is an internal `[ExcludeFromCodeGenerator]` Attribute to exclude enums from the generator. The generator uses structs, IEnumerable with yield return and no LINQ to be as efficient as possible.

While creating the PR, I noticed that there is already a PR for a source code generator. However, the approach and the implementation is different and uses the old `ISourceGenerator` which can have a bad impact in performance during development. See [here](https://andrewlock.net/exploring-dotnet-6-part-9-source-generator-updates-incremental-generators/#the-problems-with-performance-for-isourcegenerator). So, this is an alternative PR to close  #6098.


## How Has This Been Tested?
The source code of the generator and the source code builder are covered by unit tests. In addition, I tested the generated extension methods by starting the docs project. Existing tests are still working.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

P.S.: This is my first contribution to an open-source project 🥳

xC0dex committed:
SourceCodeGenerator: Added initial support for enum source code generation

Tyme-Bleyaert commented:
Closing #6135 as this one uses `IIncrementalGenerator` .

Tyme-Bleyaert reviewed (commented):
Looks good,

But have you also tested the generator with accessibility of nested classes / enums?

For example: 
```
internal class SomeClass
{
    public enum NestedEnum
    {
        None = 0,
        First,
        Second
    }
}
```

Maybe add an extra check if the containing type is less accessibly than the enum.
So you wont get this:
![image](https://user-images.githubusercontent.com/72735570/218827196-96bfa75c-2ad7-408f-8e59-e46c247478bd.png)


Other than that everything looks fine.

xC0dex committed:
SourceCodeGenerator: Added check containing access modifier and removed obsolete usings

xC0dex commented:
Hey @Tyme-Bleyaert,
thank you very much for reviewing this PR and your feedback. I added a check, if the enum is nested and if the access modifier is different. This now also covered by a unit test. In addition, I removed the old usings/references of the previous `ToDescription` method.

Tyme-Bleyaert commented on src/MudBlazor/Extensions/EnumExtensions.cs:
```diff
@@ -5,6 +5,7 @@ namespace MudBlazor.Extensions
 {
     public static class EnumExtensions
     {
+        [Obsolete("Please use custom source code generated ToDescriptionString method instead or implement a own extension method.")]
```
Also I'm not a native English speaker myself but I'm not sure "implement a own extension method" is correct.
I would think it's either "your own" or "an extension method yourself".

ScarletKuro replied:
I agree, it should be "or implement your own extension method."

xC0dex replied:
Thanks for your feedback. The obsolete message is now "_Please use the auto-generated ToDescriptionString method instead or implement your own extension method._". I hope you're fine with that.

Tyme-Bleyaert reviewed (approved):
The changes look good to me.

ScarletKuro commented:
Is there any plan on merging this or it will be delayed until .net 8?
I think this is a good work, there are performance benefits and no reflection.

xC0dex committed:
SourceCodeGenerator: Updated obsolete message and renamed project

ScarletKuro commented:
@henon sorry to tag directly, but interesting to get your opinion

henon commented:
> @henon sorry to tag directly, but interesting to get your opinion

No, it is the right thing to do. Thanks for bringing my attention on this.

ScarletKuro commented:
@Tyme-Bleyaert what I didn't notice during review, why did disable generation for `InputMode` with `ExcludeFromCodeGenerator`?

ScarletKuro commented on src/MudBlazor/Enums/InputMode.cs:
```diff
@@ -11,6 +11,7 @@ namespace MudBlazor
     /// Not supported by safari. Use Pattern to achieve special mobile keyboards in safari.
     /// https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inputmode
     /// </summary>
+    [ExcludeFromCodeGenerator]
```
What is the reason to opt-out here? Is there no `ToDescriptionString` calls on this type, if yes I kinda wonder why is there Description attribute at first place then.

Tyme-Bleyaert replied:
@ScarletKuro

That because the extension method wasn't used for that Enum anyway.
And that was the Enum that used keywords as Flags. 
So he wouldn't have to implement a lot of extra checks, just for this one Enum.

Tyme-Bleyaert replied:
The descriptions of each flag is literally a copy paste from [https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inputmode](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inputmode).
They serve no purpose other than informational I guess.

ScarletKuro replied:
Hmmm, I see. I just remember that in the other PR we specifically added support to the keyword with "@" and I thought the reason was that this type was using the extension method. Alright, then no need to change that.

henon commented:
Do we still need the old reflection-based method?

ScarletKuro commented:
> Do we still need the old reflection-based method?

There was a small discussion about it in the [old PR](https://github.com/MudBlazor/MudBlazor/pull/6135#issuecomment-1382002604). Technically it will be a breaking change then because you remove a public API that someone might use. My personal opinion is that I would remove it still.

henon commented:
Got it, we'll remove it in [The Big Break](https://github.com/MudBlazor/MudBlazor/projects/13) which is already overdue. No need to remove it in this PR just yet. Good to know we don't need it in MudBlazor any more.

henon commented:
Thank you very much @Tyme-Bleyaert, @xC0dex and @ScarletKuro !
